### PR TITLE
etag implementation

### PR DESCRIPTION
**Pull request by xiangyan99**

### CODE REVIEW

**Comment by adxsdk6** [comment]
Can one of the admins verify this patch?

File: sdk/appconfiguration/azure-appconfiguration/azure/appconfiguration/aio/azure_configuration_client_async.py [hunk 1]
```diff
@@ -188,14 +195,75 @@ def list_configuration_settings(
...
+            key_value = await self._impl.get_key_value(
+                key=key,
+                label=label,
+                if_match=if_match,
+                error_map=error_map,
+                **kwargs
+            )
+            return ConfigurationSetting._from_key_value(key_value)
+
+        if match_condition == MatchConditions.IfPresent:
```
Comment (bryevdv) [inline sdk/appconfiguration/azure-appconfiguration/azure/appconfiguration/aio/azure_configuration_client_async.py]: Copy pasta? The block above also checked `match_condition == MatchConditions.IfPresent`

File: sdk/appconfiguration/azure-appconfiguration/azure/appconfiguration/aio/azure_configuration_client_async.py [hunk 2]
```diff
@@ -188,14 +195,75 @@ def list_configuration_settings(
...
+                404: ResourceNotFoundError
+            }
+            key_value = await self._impl.get_key_value(
+                key=key,
+                label=label,
+                if_none_match=if_none_match,
+                error_map=error_map,
+                **kwargs
+            )
+            return ConfigurationSetting._from_key_value(key_value)
```
Comment (bryevdv) [inline sdk/appconfiguration/azure-appconfiguration/azure/appconfiguration/aio/azure_configuration_client_async.py]: Both the `IfModified` and `IfNotModified` blocks are identical except for `error_map` (specifically they both pass the same `if_none_match`) Is this correct?

File: sdk/appconfiguration/azure-appconfiguration/azure/appconfiguration/aio/azure_configuration_client_async.py [hunk 3]
```diff
@@ -324,18 +456,72 @@ def list_configuration_settings(
...
-        )
-        return ConfigurationSetting._from_key_value(key_value_deleted)
+        if match_condition == MatchConditions.Unconditionally:
+            key_value_deleted = await self._impl.delete_key_value(
+                key=key,
+                label=label,
+                headers=custom_headers,
+                error_map={
+                    409: ResourceReadOnlyError,
+                },
```
Comment (bryevdv) [inline sdk/appconfiguration/azure-appconfiguration/azure/appconfiguration/aio/azure_configuration_client_async.py]: Here the calls all define `error_map` at the call site. Above, `error_map` is always defined before the call. Suggest being consistent throughout a given module.

File: sdk/appconfiguration/azure-appconfiguration/azure/appconfiguration/aio/azure_configuration_client_async.py [hunk 4]
```diff
@@ -324,18 +456,72 @@ def list_configuration_settings(
...
+            key_value_deleted = await self._impl.delete_key_value(
+                key=key,
+                label=label,
+                if_none_match=if_none_match,
+                headers=custom_headers,
+                error_map={
+                    409: ResourceReadOnlyError,
+                    412: ResourceExistsError,
+                },
+            )
```
Comment (bryevdv) [inline sdk/appconfiguration/azure-appconfiguration/azure/appconfiguration/aio/azure_configuration_client_async.py]: This kind of code is repeated both across sync/async but also repeated with both across operations. What about factoring out functions that construct kwargs?

    key_value_deleted = await self._impl.delete_key_value(
        key=key,
        label=label, 
        headers=custom_headers,
        **match_args(match_condition, etag))

Where `match_args` would return something like `{'if_match': ..., 'error_map': ...}` Such a function would also be easy to put under comprehensive unit tests. The only caveat is that the error maps would need to be the union of all error maps for a given match condition, but assuming passing irrelevant entries to operations where they don't apply is harmless.

File: sdk/appconfiguration/azure-appconfiguration/azure/appconfiguration/aio/azure_configuration_client_async.py [hunk 5]
```diff
@@ -164,20 +166,25 @@ def list_configuration_settings(
...
+        # type: (str, str, str, MatchConditions, dict) -> ConfigurationSetting
 
         """Get the matched ConfigurationSetting from Azure App Configuration service
 
         :param key: key of the ConfigurationSetting
         :type key: str
         :param label: label of the ConfigurationSetting
         :type label: str
+        :param etag: check if the ConfigurationSetting is changed. Set None to skip checking etag
+        :type etag: str
```
Comment (bryevdv) [inline sdk/appconfiguration/azure-appconfiguration/azure/appconfiguration/aio/azure_configuration_client_async.py]: str or None

**Pull request merged**